app/repositories/onlinetesting/OnlineTestEvaluationRepository.scala

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2016 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package repositories.onlinetesting
+
+import model.ApplicationStatus._
+import model.Exceptions.PassMarkEvaluationNotFound
+import model.persisted._
+import model.{ ApplicationStatus, ProgressStatuses, SelectedSchemes }
+import reactivemongo.api.DB
+import reactivemongo.bson.{ BSONArray, BSONDocument, BSONObjectID }
+import repositories.{ BSONHelpers, CommonBSONDocuments, RandomSelection }
+import uk.gov.hmrc.mongo.ReactiveRepository
+import uk.gov.hmrc.mongo.json.ReactiveMongoFormats
+import model.Phase._
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+trait OnlineTestEvaluationRepository[T] extends CommonBSONDocuments with BSONHelpers with RandomSelection {
+
+  this: ReactiveRepository[_, _] =>
+
+  val phase: Phase
+
+  val evaluationApplicationStatuses: List[ApplicationStatus]
+
+  val applicationBSONReader = (doc: BSONDocument) => {
+    val applicationId = doc.getAs[String]("applicationId").get
+    val applicationStatus = doc.getAs[ApplicationStatus]("applicationStatus").get
+    val isGis = doc.getAs[BSONDocument]("assistance-details").exists(_.getAs[Boolean]("guaranteedInterview").contains(true))
+    val preferences = doc.getAs[SelectedSchemes]("scheme-preferences").get
+    val applicationBuilder: (List[CubiksTest], Option[PassmarkEvaluation]) => ApplicationReadyForEvaluation =
+      ApplicationReadyForEvaluation(applicationId, applicationStatus, isGis, _, _, preferences)
+    applicationBuilder.curried
+  }
```

Comment:
the use the of .curried here is fairly clever, but I would argue that some readability has been lost. For example, the parameters that are to be applied later are represented as underscores, so we don't actually know what is missing. I would also argue that the doc is the think that varies most, so should perhaps be the last param. See if you think this is any more readable/explicit?
```
  def applicationBSONReaderBuilder(tests: List[CubiksTest], pm: Option[PassmarkEvaluation])(doc: BSONDocument) = {
    val applicationId = doc.getAs[String]("applicationId").get
    val applicationStatus = doc.getAs[ApplicationStatus]("applicationStatus").get
    val isGis = doc.getAs[BSONDocument]("assistance-details").exists(_.getAs[Boolean]("guaranteedInterview").contains(true))
    val preferences = doc.getAs[SelectedSchemes]("scheme-preferences").get
    ApplicationReadyForEvaluation(applicationId, applicationStatus, isGis, tests, pm, preferences)
  }
```